### DISCUSSION TITLE
compilerOptions.isCustomElement in Quasar

### DISCUSSION THREAD

**Discussion by IvanWala** [quasarframework/quasar#11752] [General - Components / Directives / etc] [answered]
Hi I try to use lottie in Quasar.

The only way I can render a lottie animation in vue 3 is use the LottieFiles web player (because any of the lottie's vue wrappers doesn't is ready for vue 3). 

https://lottiefiles.com/web-player

You can read thomalex with the same problem with vue 3 and lottie here.  

https://stackoverflow.com/questions/69582182/vue-3-lottie-player-failed-to-resolve-component

The problem is it needs to edit compilerOptions.isCustomElement in Quasar but I doesn't find any documentation for do it.

**Comment by pdanpdan** [comment]
Search for compilerOptions here https://quasar.dev/quasar-cli/quasar-conf-js#property-build

**Comment by IvanWala** [comment]
Hi @pdanpdan. Thanks.

Here doesn't explain how to use it in quasar. The documentation is incomplete and i doesn't find nothing over this board.

**Comment by pdanpdan** [comment]
How is it incomplete? What is not specified (related to how to pass the option to vue-loader)?

**Comment by IvanWala** [comment]
Hi & thanks again @pdanpdan 

The documentation doesn't explain how to edit the vueLoaderOptions. It only says you can edit it here but not how. 

I try it like this: 

```
      vueLoaderOptions: {
        isCustomElement: tag => tag === 'lottie-player'
      },
```

But this doesn't work.

**Comment by pdanpdan** [comment, marked as answer]
you missed one level in the object (compilerOptions)

**Reply by IvanWala** [reply]
Thanks. That is!!.

```
      vueLoaderOptions: {
        compilerOptions: {
          isCustomElement: tag => tag === 'lottie-player'          
        }
      },
```

**Reply by ssuess** [reply]
WHERE exactly does this block reside? In `quasar.config.js` I have tried putting it at top level (right next to `boot: []`) also under `build: {`, and also under `vitePlugins` and can't get it to work anywhere.

**Reply by yusufkandemir** [reply]
For Webpack it's under `quasar.config.js > build`: https://quasar.dev/quasar-cli-webpack/quasar-config-js#property-build

For Vite the configuration is a bit different: https://github.com/quasarframework/quasar/discussions/11752#discussioncomment-3078068

**Reply by madLinux7** [reply]
> For Webpack it's under `quasar.config.js > build`: https://quasar.dev/quasar-cli-webpack/quasar-config-js#property-build
> 
> For Vite the configuration is a bit different: [#11752 (comment)](https://github.com/quasarframework/quasar/discussions/11752#discussioncomment-3078068)

this part is important! Thanks!!

**Comment by pdanpdan** [comment]
Reply to your deleted comment:

> No this doesn't works also:
> 
>       vueLoaderOptions: {
>         compilerOptions: {
>           isCustomElement: tag => tag === 'lottie-player'          
>         }
>       },
> runtime-core.esm-bundler.js?9e79:6568 [Vue warn]: Failed to resolve component: lottie-vue-player
> If this is a native custom element, make sure to exclude it from component resolution via compilerOptions.isCustomElement. 
>   at <LottieSpinner class="q-loading__spinner" color="" size=80 > 
>   at <BaseTransition onAfterLeave=fn<onAfterLeave> appear=true persisted=false  ... > 
>   at <Transition name="q-transition--fade" appear=true onAfterLeave=fn<onAfterLeave> > 
>   at <QLoading>

Because it looks like it's `lottie-vue-player` not `lottie-player`

**Reply by IvanWala** [reply]
> Reply to your deleted comment:
> 
> > No this doesn't works also:
> > ```
> >   vueLoaderOptions: {
> >     compilerOptions: {
> >       isCustomElement: tag => tag === 'lottie-player'          
> >     }
> >   },
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > runtime-core.esm-bundler.js?9e79:6568 [Vue warn]: Failed to resolve component: lottie-vue-player
> > If this is a native custom element, make sure to exclude it from component resolution via compilerOptions.isCustomElement.
> > at 
> > at <BaseTransition onAfterLeave=fn appear=true persisted=false  ... >
> > at <Transition name="q-transition--fade" appear=true onAfterLeave=fn >
> > at
> 
> Because it looks like it's `lottie-vue-player` not `lottie-player`

I edited my previus comment some seconds before this answer.

**Comment by ssuess** [comment]
From the docs, it seems like this should be correct:
```
build: {
      vueCompiler: true,
      vueLoaderOptions: {
        compilerOptions: {
          isCustomElement: tag => tag.startsWith('mobile-menu')
        }
      },
 ```
but the warnings remain...
btw, this worked fine in webpack:
```
chainWebpack (chain, { isServer, isClient }) {
        chain.module
        .rule('vue')
        .use('vue-loader')
        .tap(options => {
          options.compilerOptions = {
            ...options.compilerOptions,
            isCustomElement: tag => tag.startsWith('mobile-menu')
          }
          return options
        })
  ```

**Comment by ssuess** [comment]
Have also tried this in vitePlugins:
```
['@vitejs/plugin-vue', {
          template: {
            compilerOptions: {
              isCustomElement: (tag) => tag.includes('mobile-menu')
            }
          }
        }]
```
which results in `[plugin:vite:vue] Illegal tag name. Use '&lt;' to print '<'.`
and tried a boot file:
```
import { boot } from 'quasar/wrappers'

export default boot(({ app }) => {

  app.config.compilerOptions.isCustomElement = (tag) => {
    return tag.startsWith('mobile-menu')
  }
})
```
Which ALSO failed. 
also tried:
```
extendViteConf (viteConf, { isClient, isServer }) {
        viteConf.plugins.push(
          require('@vitejs/plugin-vue')({
            template: {
              compilerOptions: {
                isCustomElement: (tag) => tag === 'mobile-menu'
              }
            }
          }),
        )
      },
```

Which of course, ALSO FAILED.
Can anyone please explain to me how to get quasar to properly ignore custom elements when using vite? I am just about to return to webpack, which although slower, works.

**Reply by yusufkandemir** [reply]
https://github.com/quasarframework/quasar/discussions/13862#discussioncomment-3078041

**Comment by yusufkandemir** [comment]
## For `@quasar/app-vite`

https://vuejs.org/guide/extras/web-components.html#example-vite-config
https://quasar.dev/quasar-cli-vite/quasar-config-js#build

```js
// quasar.config.js

build: {
  viteVuePluginOptions: {
    template: {
      compilerOptions: {
        isCustomElement: (tag) => tag.startsWith('my-')
      }
    }
  }
}
```
